src/bindings/python/flux/job.py

```diff
@@ -86,6 +86,88 @@ def job_kvs_guest(flux_handle, jobid):
     return flux.kvs.get_dir(flux_handle, kvs_key)
 
 
+def id_parse(jobid_str):
+    """
+    returns: An integer jobid
+    :rtype int
+    """
+    jobid = ffi.new("flux_jobid_t[1]")
+    RAW.id_parse(jobid_str, jobid)
+    return int(jobid[0])
+
+
+def id_encode(jobid, encoding="f58"):
+    """
+    returns: Jobid encoded in encoding
+    :rtype str
+    """
+    buflen = 128
+    buf = ffi.new("char[]", buflen)
+    RAW.id_encode(int(jobid), encoding, buf, buflen)
+    return ffi.string(buf, buflen).decode("utf-8")
+
+
+class JobID(int):
+    """Class used to represent a Flux JOBID
+
+    JobID is a subclass of `int`, so may be used in place of integer.
```

Comment:
Minor comment, but I'm not sure I follow the subclassing of `int`.  What scenarios do you have in mind where you would want to use a JobID in place of an integer?  Does this help with the auto-generated bindings which are expecting an int for the `id_parse` and `id_encode` functions?

My main concern is that the `__add__`, `__sub__`, `__mul__`, and `__div__` methods that are inherited from `int` don't really make sense in the context of a FLUID.  I suspect the only int-like methods we care about are ordering related, and we can get that with very little code using [functools `total_ordering` decorator](https://docs.python.org/3.6/library/functools.html#functools.total_ordering).

Reply:
PS - I was originally thinking that `int` was the class we were advised not to subclass in the python course (for reasons that escape me right now), but it was actually `list` that we were warned about with [`UserList`](https://docs.python.org/3.8/library/collections.html#userlist-objects) being the preferred thing to subclass.  So I think my "spidey sense was tingling" over an unrelated data type.

Reply:
My main concern was that JOBIDs were treated explicitly as `int` already in the code, and I just wanted to extend that to allow the other representations, instead of adding a swath of functions to convert to/from `int` and the other representations. We were already treating JOBIDs as integers and the `int` methods didn't seem to get in the way, so I actually thought this was the least change for the least effort.

However, I'm not a good Python programmer (clearly), what would be proper to do here?

Reply:
Note that this also follows the C API, where we have `typedef uint64_t fluid_t`.

Reply:
Sorry, one more thing. I remembered what I was worried about.

`flux-jobs` currently stores JOBIDs as `int`, and this value is passed to the `{id}` format key as `int` and therefore allows all integer format specifiers, presentations, etc. Therefore the `JobID` class needs to present itself as an `int`, but allow the other representations as properties so we can use `{id.f58}` `{id.words}` etc.  Is there a better way to accomplish this than subclassing `int`?

Reply:
> My main concern was that JOBIDs were treated explicitly as int already in the code, and I just wanted to extend that to allow the other representations, instead of adding a swath of functions to convert to/from int and the other forms. 

Gotcha! That makes total sense to me.  My concern about the arithmetic methods is relatively minor, and I see how adding calls to `int()` all over the place would be a lot, especially if we miss something that is behind a conditional (darn dynamically-typed languages!).

> flux-jobs currently stores JOBIDs as int, and this value is passed to the {id} format key as int and therefore allows all integer format specifiers, presentations, etc.

That makes a ton of sense too.  I don't know of an easy way to support all of the `int` format specifiers other than subclassing.  The hard way would be to use/re-implement the `int`'s `__format__` method or implement a `.int()` method and then do `{id.int:...}` in all the different formats that we do.

> However, I'm not a good Python programmer

This is the only thing you said that I disagree with :smile: